Add job shell staging plugin

SteVwonder opened the pull request:
This PR demonstrates the end-to-end capability of staging to multiple storage tiers within a single jobspec.  It includes sharness tests of the job-shell in standalone mode as well as end-to-end job submission tests.  These tests verify that you can stage to node-local storage, cluster-local storage, and to both simultaneously.  In the node-local case, every job shell rank that has a storage unit assigned to it stages into the storage.  In the cluster-local case, only the rank 0 job shell stages the data.

Other additions:

- Adds the nlohmann json library for more native C++ json handling
- Adds support for alternate writer formats for `find` (#736)

dongahn commented:
Awesome! This is a significant addition to our L2 milestone this year, @SteVwonder! We will surely circle back to this next fiscal year to land this.

SteVwonder commented:
Now that I've completed the refactor I wanted to do, and Travis is happy, I think this PR is finally ready for review.

dongahn commented:
@SteVwonder: so that you know. This is my next priority.

dongahn commented on t/data/resource/jobspecs/advanced/test006.yaml:
```diff
@@ -0,0 +1,39 @@
+version: 1
```
This would not be version 1.

dongahn commented on t/t3006-resource-advanced.t:
```diff
@@ -64,4 +64,15 @@ test_expect_success LONGTEST "${test004_desc}" '
 '
 
+cmds005="${cmd_dir}/cmds05.in"
+test005_desc="Resource labels in jobspec carrythrough to JGF R"
```
```suggestion
test005_desc="Resource labels in jobspec carry through to JGF R"
```

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+
+        std::string granularity = storage_attr["granularity"].get<std::string>();
+        if ((granularity != "node") && (granularity != "job")) {
+            shell_log_error ("Storage entry contains a granularity not 'job' or 'node' (%s)", granularity.c_str());
```
line too long?

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+    try {
+        info = json::parse(info_cstr);
+    } catch (json::parse_error& e) {
+        shell_log_error ("Error parsing shell \"info\" json at byte %zu. %d: %s", e.byte, e.id, e.what());
```
line too long?

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+
+    JobspecState jobspec_state = validate_jobspec (jobspec);
+    switch (jobspec_state) {
+    case JobspecState::invalid: errno=EINVAL; return -1;
+    case JobspecState::valid_no_staging: return 0;
+    case JobspecState::valid_staging: break;
```
Probably more conventional and readable to have each statement written on its own line.

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+namespace fs = boost::filesystem;
+using json = nlohmann::json;
+
+enum JobspecState { invalid, valid_no_staging, valid_staging};
```
We have been using `enum class` instead of plain `enum`:
https://github.com/flux-framework/flux-sched/blob/master/resource/schema/resource_data.hpp#L48

We have not been using camel back style. So my suggestion would be `enum class job_spec_state_t`

The fields would be better with capital: e.g., `INVALID`.

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+
+enum JobspecState { invalid, valid_no_staging, valid_staging};
+
+static JobspecState validate_jobspec (json jobspec)
```
Did you mean to pass jobspec by value?

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+            return JobspecState::invalid;
+        }
+
+        std::string granularity = storage_attr["granularity"].get<std::string>();
```
Did you mean to assign by value?

dongahn commented on src/shell/data-staging.cpp:
```diff
@@ -0,0 +1,243 @@
+    std::map<std::string, json> storage_resources;
+    json jgf;
+    try {
+        jgf = R["scheduling"]["graph"];
```
There are lots of uses of `operator[]`. I don't know what the underlying container this new json library is using. But there are various side effects with `operator[]` if this is `std::map` type. For example, if the key is non-existent, this puts a temporary value object under that key. 

The rule of thumb has been to use this method only when the key is guaranteed to exist. Just wanted to leave a note in case this has been your assumption too...

SteVwonder replied:
Good call.  It looks like that behavior is indeed used in this json library to make it STL-like.  I switched the accesses over to [`.at()`](https://nlohmann.github.io/json/api/basic_json/at/)

dongahn replied:
Yeah, that sounds good. `at` can raise an exception (out of range) so please make sure you catch this exception. I don't know if this applies here, but having constness to methods that don't change the object can also be helpful to get compile time checks...

dongahn commented on t/Makefile.am:
```diff
@@ -81,7 +81,8 @@ TESTS = \
     t6001-match-formats.t \
-    t6002-graph-hwloc.t
+    t6002-graph-hwloc.t \
+	t7000-shell-datastaging.t
```
Seems a tab is used instead of whitespaces or similar. The indentation looks incorrect with Github renderer.

dongahn commented on t/data/shell/cluster-local/jobspec.yaml:
```diff
@@ -0,0 +1,38 @@
+version: 1
```
Version 9999 instead

SteVwonder replied:
Unfortunately, the job shell still requires that `version` == `1`.  I missed that change in my other PRs to flux-core that enabled the end-to-end support.  I opened an issue (https://github.com/flux-framework/flux-core/issues/3310).  If there are no objections from the flux-core team to the suggestion, I'll allow `version` == `2` and change these to `version: 2`.

SteVwonder replied:
On the team call, we agreed to let this go as is, and bump the version once the issue in flux-core is closed.  I opened an issue in flux-sched to track this: https://github.com/flux-framework/flux-sched/issues/770

dongahn commented on t/data/shell/combined/jobspec.yaml:
```diff
@@ -0,0 +1,46 @@
+version: 1
```
Version

dongahn commented on t/data/shell/cluster-local/jobspec.yaml:
```diff
@@ -0,0 +1,38 @@
+        verbose: 1
+  user:
+    data-staging:
+      - test: true
```
For my own edification, what is the `test` key?

SteVwonder replied:
The test key signals to the data-staging shell plugin that it shouldn't actually stage any files and instead just [log a message with the file source and destination](https://github.com/SteVwonder/flux-sched/blob/job-shell-staging-plugin/src/shell/data-staging.cpp#L170).  The tests in `t700-shell-datastaging.t` then verify that the log contains the correct source and destination.

dongahn commented on t/data/shell/node-local/jobspec.yaml:
```diff
@@ -0,0 +1,37 @@
+version: 1
```
Version

dongahn commented on t/t7000-shell-datastaging.t:
```diff
@@ -0,0 +1,141 @@
+
+shell_basepath=`readlink -e ${SHARNESS_TEST_SRCDIR}/data/shell`
+shell_plugin_path="$(readlink -e ${SHARNESS_BUILD_DIRECTORY}/src/shell/.libs)"
+grug_basepath=`readlink -e ${SHARNESS_TEST_SRCDIR}/data/resource/grugs/`
```
This variable is unused.

dongahn reviewed (changes requested):
@SteVwonder: Thank you for the work! This is a great way to be prepare for future multi-tired storage support. 

I just had my first pass. As you requested, I looked more closely on your C++ codes added to the existing resource infrastructure. There are things like memory copy semantics I want you to look at. 

I had some cursory look at the jobshell plugin src codes and added some comments. There are some cleanup that would be good to have.

Once that's done, I will have another pass.

dongahn commented on resource/traversers/dfu_impl.cpp:
```diff
@@ -602,6 +602,13 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
+    for (auto &resource: resources) {
+        if ((resource.type == (*m_graph)[u].type) &&
+            (!resource.label.empty())) {
+            (*m_graph)[u].idata.ephemeral.insert (m_best_k_cnt, "label", resource.label);
```
Oops. Meant to leave a note. This would need an error check.

SteVwonder commented:
Thanks @dongahn! All your suggestions seem doable in the next couple of days, so I will attempt to have a new version ready for review by the end of the week.

Excellent catch on the missing comparison in the tests.  I hope I just accidentally lost that line in a rebase somewhere.  🤦 

And thanks for taking an extra look at the C++ copy-relevant stuff.  That was exactly the kind of thing I was worried I was doing incorrectly without realizing it.

SteVwonder commented on t/t1015-find-format.t:
```diff
@@ -0,0 +1,105 @@
+EOF
+    ${query} -d -L ${excl_1N1B}/0.xml -f hwloc -F simple -S CA -P high \
+             -W node,core -t simple.query.raw < find.cmd &&
+    strip_info_lines < simple.query.raw > simple.query.out
```
Yeah, this is confusing.

Rather than add a bunch of new json/txt files to compare the output of `flux ion-resource find` against, I instead opted to generate the "ground truth" find files using `resource-query`, which is what is happening in these first few tests.

Depending on what you think is best, I could:

- just add a comment to the tests outlining what is happening and why the first set of tests don't have a comparison being performed
- add explicit "ground truth" files to the repo to compare against (in which case the first few tests could serve to test that `resource-query` outputs the correct find format too)

SteVwonder commented on resource/writers/match_writers.cpp:
```diff
@@ -215,11 +215,13 @@ int jgf_match_writers_t::emit (std::stringstream &out)
     int rc = 0;
     json_t *o = NULL;
     json_t *b = NULL;
+    auto ephemeral = g[u].idata.ephemeral;
+    auto eph_map = ephemeral.to_map (epoch);
```
Good catch.  It looks like I'm bumping up against the fact that `g` is passed in as `const` to `emit_vtx`:

```
writers/match_writers.cpp: In member function ‘virtual int Flux::resource_model::jgf_match_writers_t::emit_vtx(const string&, const f_resource_graph_t&, const vtx_t&, unsigned int, bool, uint64_t)’:
writers/match_writers.cpp:223:55: error: passing ‘const Flux::resource_model::ephemeral_t’ as ‘this’ argument discards qualifiers [-fpermissive]
     auto &eph_map = g[u].idata.ephemeral.to_map (epoch);
                                                       ^
In file included from ../resource/schema/infra_data.hpp:29:0,
                 from ../resource/schema/resource_data.hpp:35,
                 from writers/match_writers.cpp:27:
../resource/schema/ephemeral.hpp:39:41: note:   in call to ‘std::map<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >& Flux::resource_model::ephemeral_t::to_map(uint64_t)’
     std::map<std::string, std::string>& to_map (uint64_t epoch);
```

If we were just attempting to convert the ephemeral container into a map with non-stale data, then const would be fine (although require some extra copying).  The sticking point is the call to `ephemeral_t::clear` further down in this function.

@dongahn: how would you recommend proceeding?

One option that I see would be to remove the `const` qualifier from `g` in `emit_vtx`, but that doesn't seem like a great option.  Are we OK with `emit_vtx` no longer be idempotent?

Another might be a `post_emit` or `vtx_cleanup` method within `dfu_impl.cpp` which is called after the `emit_vtx` and wouldn't have the `const` qualifier on `g` (or `needs`, `exclusive`, or `epoch`).  It would also simplify the cleanup of the vertices for other match writers (which in writing this, I just realized I did not attempt any `clear`ing of the ephemeral_t for non-jqf writers like rv1).

dongahn replied:
I think `emit_vtx` remains to be `const` is preferred since a `emit` method typically implies read only. Is it possible to update `g` after `emit` as part of the infrastructure code (`dfu_impl.cpp`)? Or do you really need another callback?

dongahn replied:
I don't know if it is possible,  but if "epoch" and such can also be hidden into the infrastructure code without having to expose it to the callback, that'd be even better. At least that's how I hide such implementation complex for other similar logic like coloring and state cleaning.

SteVwonder replied:
I refactored things in [`0820b6c` (#743)](https://github.com/flux-framework/flux-sched/pull/743/commits/0820b6cf001d03005af3357c6671ff4e4b4afcfd).  Most of the commit is undoing what was done in [`3dc3989` (#743)](https://github.com/flux-framework/flux-sched/pull/743/commits/3dc3989df79fae0c0ed022ee553cce874c993152). I dropped the `epoch` argument from the `emit_vtx` callbacks, and just wrapped the one call to `emit_vtx` in `dom_find_dfv` with calls to `ephemeral.check_and_clear_if_stale` and `ephemeral.clear` (plus lots of comments).  I think it actually ends up being nicer/cleaner than what was there before due to what you mention:

> if "epoch" and such can also be hidden into the infrastructure code without having to expose it to the callback, that'd be even better. At least that's how I hide such implementation complex for other similar logic like coloring and state cleaning.

Let me know if you think this refactor is ultimately the right way to go.  I'm happy to redo it a different way.

SteVwonder replied:
@milroy: in your next review, I'd like to call your attention in particular to the refactor described in the above comment.

milroy commented on resource/schema/ephemeral.cpp:
```diff
@@ -0,0 +1,84 @@
+
+bool ephemeral_t::check_and_clear_if_stale (uint64_t epoch)
+{
+    if (this->epoch < epoch) {
```
Is it possible for the current epoch passed to `check_and_clear_if_stale` to be less than one second after `this->epoch` is set (i.e. the data is actually stale and should fail this check)?

milroy replied:
I'm answering my own question based on a conversation with @SteVwonder.  I interpreted `epoch` to be a timestamp (i.e. Unix epoch).  However, `epoch` is a counter, which will not cause the issue in my question.

milroy reviewed (commented):
This looks really good overall!  Most of my comments are nits, but I do have a question about a potential corner case for testing for stale data with `epoch`.

SteVwonder commented on resource/Makefile.am:
```diff
@@ -25,6 +25,7 @@ libresource_la_SOURCES = \
     schema/infra_data.cpp \
```
I used the backticks with the thought that Github would render them as literal strings via its markdown parser (see below for how it renders on GH).  I'm happy to change them to `""` if you think that is the better format outside of GH:

<img width="532" alt="Screen Shot 2020-11-05 at 12 09 05 PM" src="https://user-images.githubusercontent.com/874707/98291304-c3aa3c80-1f5f-11eb-9ecf-1c690192ffc7.png">

SteVwonder commented:
Thanks @dongahn and @milroy for the great comments!  This is ready for another round of reviews.  There is a stack of `fixup!` commits at the top of the commit stack that contain the changes since the last review round.

milroy commented on resource/traversers/dfu_impl.cpp:
```diff
@@ -606,7 +606,15 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
         if ((resource.type == (*m_graph)[u].type) &&
             (!resource.label.empty())) {
-            (*m_graph)[u].idata.ephemeral.insert (m_best_k_cnt, "label", resource.label);
+            if ((*m_graph)[u].idata.ephemeral.insert (m_best_k_cnt,
+                                                      "label",
+                                                      resource.label) < 0) {
+                m_err_msg += "dom_dfv: inserting label into ephemeral failed.\n";
+                m_err_msg += strerror (errno);
+                m_err_msg += ".\n";
+                rc = -1;
```
The pattern throughout `dfu_impl.cpp` is to set the return code in the condition.  Unfortunately, my suggestion extends beyond 80 characters on one line, and I'm not sure the best way to split such a short line. 
```suggestion
            if ( (rc = (*m_graph)[u].idata.ephemeral.insert (m_best_k_cnt,
                                                             "label",
                                                             resource.label)) < 0) {
                m_err_msg += "dom_dfv: inserting label into ephemeral failed.\n";
                m_err_msg += strerror (errno);
                m_err_msg += ".\n";
```

SteVwonder replied:
I ended up just splitting is across two lines (and dropping the rc = -1):

```
           rc = (*m_graph)[u].idata.ephemeral.insert (m_best_k_cnt,
                                                       "label",
                                                       resource.label);
            if (rc < 0) {
```

milroy commented on src/shell/data-staging.cpp:
```diff
@@ -23,97 +23,116 @@ extern "C" {
-                return JobspecState::invalid;
+            if (!storage_attr.contains (req_attr)) {
+                shell_log_error ("Storage entry missing %s key",
+                                 req_attr.c_str());
```
Nit: space before parenthesis.
```suggestion
                                 req_attr.c_str ());
```

milroy commented on src/shell/data-staging.cpp:
```diff
@@ -23,97 +23,116 @@ extern "C" {
-            return JobspecState::invalid;
+            shell_log_error ("Storage entry contains a granularity not "
+                             "equal to 'job' or 'node' (%s)",
+                             granularity.c_str());
```
Nit: space before parenthesis.
```suggestion
                             granularity.c_str ());
```

milroy commented on src/shell/data-staging.cpp:
```diff
@@ -23,97 +23,116 @@ extern "C" {
     json jgf;
     try {
-        jgf = R["scheduling"]["graph"];
+        jgf = R.at("/scheduling/graph"_json_pointer);
```
Nit: space before parenthesis.
```suggestion
        jgf = R.at ("/scheduling/graph"_json_pointer);
```

milroy commented on src/shell/data-staging.cpp:
```diff
@@ -197,30 +220,36 @@ static int main_init (flux_plugin_t *p)
-        R = info["R"];
+        rank = info.at ("rank").get<int>();
+        jobspec = info.at ("jobspec");
+        R = info.at("R");
```
Nit: space before parenthesis.
```suggestion
        R = info.at ("R");
```

milroy commented on resource/schema/ephemeral.cpp:
```diff
@@ -71,6 +71,11 @@ boost::optional<std::string> ephemeral_t::get (uint64_t epoch, const std::string
 const std::map<std::string, std::string>& ephemeral_t::to_map (uint64_t epoch)
 {
     check_and_clear_if_stale (epoch);
+    return this->to_map();
```
Nit: spacing
```suggestion
    return this->to_map ();
```

milroy reviewed (changes requested):
This looks great @SteVwonder and just needs a few minor style changes.

dongahn commented:
FWIW, @SteVwonder and I had a little on-line code review for stale data deallocation scheme (as this was one that he need me to take a look) and all look great to me.

dongahn commented:
This is the last PR that needs to land before we can tag our next release, correct?

SteVwonder commented:
Yeah. This PR and then the release notes PR, which is already up.  I'm going to grab some dinner and then I'll be back to wrap this PR up. Should be relatively painless. 

Thanks @milroy for the thorough review!

dongahn commented:
Cool! I have to say this release cycle was pretty productive! Thank you everyone!

SteVwonder commented:
Just integrated @milroy's formatting changes and squashed down the `fixup!` commits.

@milroy or @dongahn: if you are happy with the PR, could you approve and toss the MWP label on?  Thanks again for all of the feedback.

SteVwonder commented:
Darn.  Mergify won't go if there is still a reviewer requesting changes. @milroy, could you change your review to `accept` to appease mergify?  Sorry for bothering you on the weekend.

dongahn commented:
In this case, I believe it makes sense to manually merge this.

milroy commented:
Sorry @SteVwonder to forget about the changing my review to `accept`.  I'll remember next time!
